### PR TITLE
Introduce Forwarded/X-Forwarded header config

### PR DESCRIPTION
**Pull request by chamil321**
- If the first interface to the client will set Forwarded header with for, by, host and proto parameters.
- If the Forwarded header is available in the inbound request, update it.
- If the X-Forwarded-- headers are available in the inbound request, update them.
> User set config mode - "Transition"
- If the Forwarded header is available, update it.
- If the X-Forwarded-- headers are available in the inbound request, transform to Forwarded.

## Automation tests
 - Unit tests 
   > Added
 - Integration tests
   > -

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

### CODE REVIEW

File: components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/config/ForwardedExtensionConfig.java [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package org.wso2.transport.http.netty.config;
```
Comment (anupama-pathirage) [inline components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/config/ForwardedExtensionConfig.java]: Missing license header

File: components/org.wso2.transport.http.netty/src/test/java/org/wso2/transport/http/netty/unitfunction/ForwardedTestCase.java [hunk 2]
```diff
@@ -0,0 +1,192 @@
+package org.wso2.transport.http.netty.unitfunction;
```
Comment (irunika) [inline components/org.wso2.transport.http.netty/src/test/java/org/wso2/transport/http/netty/unitfunction/ForwardedTestCase.java]: Missing license header

Comment (irunika) [inline components/org.wso2.transport.http.netty/src/test/java/org/wso2/transport/http/netty/unitfunction/ForwardedTestCase.java]: License header is missing

File: components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/contractimpl/HttpClientConnectorImpl.java [hunk 3]
```diff
@@ -53,6 +54,7 @@
     private String httpVersion;
     private ChunkConfig chunkConfig;
     private boolean keepAlive;
+    private ForwardedExtensionConfig forwardedExtensionConfig;
 
     public HttpClientConnectorImpl(ConnectionManager connectionManager, SenderConfiguration senderConfiguration) {
```
Comment (irunika) [inline components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/contractimpl/HttpClientConnectorImpl.java]: Is it possible to change the class name to DefaultHttpClientConnector in this PR?

File: components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/sender/ForwardedHeaderUpdater.java [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+     *
+     * @see <a href="https://tools.ietf.org/html/rfc7239">rfc7239</a>
+     */
+    public void setForwardedHeader() {
+        StringBuilder headerValue = new StringBuilder();
+        if (forwardedHeader == null) {
+            Object remoteAddressProperty = httpOutboundRequest.getProperty(Constants.REMOTE_ADDRESS);
+            if (remoteAddressProperty != null) {
+                String remoteAddress = ((InetSocketAddress) remoteAddressProperty).getAddress().getHostAddress();
+                headerValue.append(FOR + resolveIP(remoteAddress) + "; ");
```
Comment (Manuri) [inline components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/sender/ForwardedHeaderUpdater.java]: We can make semi-colon, colon, comma etc constants as they are used in many places.

File: components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/sender/channel/TargetChannel.java [hunk 5]
```diff
@@ -266,4 +269,26 @@ private void writeOutboundRequestHeaders(HTTPCarbonMessage httpOutboundRequest)
...
+        }
+        if (headerUpdater.isTypeXForwarded()) {
+            if (forwardedConfig == ForwardedExtensionConfig.ENABLE) {
+                headerUpdater.setDefactoForwardedHeaders();
+                return;
+            }
+            headerUpdater.transformAndSetForwardedHeader();
+            return;
+        }
+        log.warn("Both Forwarded and X-Forwarded-- headers are present. Hence updating only the forwarded header");
```
Comment (Manuri) [inline components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/sender/channel/TargetChannel.java]: Here, do we consider both types of headers are present if both headerUpdater.isTypeForwarded() and headerUpdater.isTypeXForwarded() conditions are false?

File: components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/config/ForwardedExtensionConfig.java (hunk 1 above)
Comment (chamil321) [reply components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/config/ForwardedExtensionConfig.java]: Ack

File: components/org.wso2.transport.http.netty/src/test/java/org/wso2/transport/http/netty/unitfunction/ForwardedTestCase.java (hunk 2 above)
Comment (chamil321) [reply components/org.wso2.transport.http.netty/src/test/java/org/wso2/transport/http/netty/unitfunction/ForwardedTestCase.java]: Ack

File: components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/contractimpl/HttpClientConnectorImpl.java (hunk 3 above)
Comment (chamil321) [reply components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/contractimpl/HttpClientConnectorImpl.java]: Ack

File: components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/sender/channel/TargetChannel.java (hunk 5 above)
Comment (chamil321) [reply components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/sender/channel/TargetChannel.java]: I changed function name to headerUpdater.isForwardedHeaderRequired(). This function tells whether this request needs a forwarded header to be set or not. 
There can be a situation where both Forwarded and X-Forwarded headers are present in the inbound request. In such case, ballerina only update/set Forwarded header as it is the standard header among them.
So we log a warning and set Forwarded header.

File: components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/common/Constants.java [hunk 6]
```diff
@@ -111,6 +111,11 @@
     // TODO: Move string constants for HTTP header values to their own class
     public static final String CONNECTION_KEEP_ALIVE = "keep-alive";
     public static final String CONNECTION_CLOSE = "close";
+    public static final String FORWARDED = "Forwarded";
+    public static final String X_FORWARDED_FOR = "X-Forwarded-For";
+    public static final String X_FORWARDED_BY = "X-Forwarded-By";
+    public static final String X_FORWARDED_HOST = "X-Forwarded-Host";
+    public static final String X_FORWARDED_PROTO = "X-Forwarded-Proto";
```
Comment (shafreenAnfar) [inline components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/common/Constants.java]: Shall we make it lowercase ?

**Pull request merged**